Add db connection pool support

arturmkr opened the pull request:
### Summary
This PR introduces optional JDBC connection pooling using HikariCP in `JdbcConnectionManager`, solves issue #377 

### Details
- Added an optional field `maxPoolSize` to `DataStoreConfiguration`.
- When `maxPoolSize` is configured (non-null and > 0):
  - `JdbcConnectionManager` uses a Hikari connection pool for routing-group database connections.
- When `maxPoolSize` is not set (null):
  - Connections are created directly (no pooling).

### Tests
Added `TestJdbcConnectionManagerPool`:
1. **blocksWhenExceedingMaxPoolSize**
   - Verifies that when a pool is enabled, new connections block after the pool limit.
2. **doesNotBlockWhenMaxPoolSizeIsNull**
   - Verifies that when the pool is disabled, connections are created directly and do not block.

### Why
This allows controlling connection usage and improving DB efficiency under load, while keeping the old behavior as the default (no pooling).

## Summary by Sourcery

Introduce optional JDBC connection pooling using HikariCP controlled by a new maxPoolSize setting, while preserving default direct connections, and add tests to verify pooling behavior

New Features:
- Enable JDBC connection pooling via HikariCP when maxPoolSize > 0

Enhancements:
- Add maxPoolSize field with constructors, getter, and setter to DataStoreConfiguration
- Update JdbcConnectionManager#getJdbi to use a HikariCP DataSource when pooling is enabled

Build:
- Add HikariCP dependency to the project

Tests:
- Introduce TestJdbcConnectionManagerPool with tests for pool blocking and no-pool scenarios

rdsarvar reviewed (commented):
small nitpick on configurability but open to @mosabua et al opinions

arturmkr commented on gateway-ha/src/main/java/io/trino/gateway/ha/config/DataStoreConfiguration.java:
```diff
@@ -21,15 +21,32 @@ public class DataStoreConfiguration
     private String driver;
     private Integer queryHistoryHoursRetention = 4;
     private boolean runMigrationsEnabled = true;
+    private Integer maxPoolSize;
```
Thanks for the suggestion! I think exposing all HikariCP parameters would add unnecessary complexity for this use case as Hikari’s default settings are already well-optimized and sufficient.
For now I kept it minimal with maxPoolSize, since it’s the only setting that typically needs adjustment.

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/ha/persistence/JdbcConnectionManager.java:
```diff
@@ -59,7 +66,18 @@ public Jdbi getJdbi(@Nullable String routingGroupDatabase)
         }
 
-        return Jdbi.create(buildJdbcUrl(routingGroupDatabase), configuration.getUser(), configuration.getPassword())
+        Integer maxPoolSize = configuration.getMaxPoolSize();
```
Disagreed, we should expose only important properties.

kbhatianr replied:
My comment wasn't what we should expose / not expose (well, I did put my opinion that imo we should expose the entire Hikari config), but rather that we shouldn't have individual properties that we pass around - instead, creating a container object for this allows easier extensibility in the future when we may want additional properties.

Chaho12 reviewed (commented):
Can you update docs here? regarding pool size? I  haven't used HikariCP so am not too famliar with this library.
https://github.com/trinodb/trino-gateway/blob/3a1553a401a430fa7f4ce3d00791c6e3f0af4f81/docs/installation.md?plain=1#L54-L63

A reference to Hikari might be good too, suggesting 10 is suitable number for decent performance for db connection pool support.
https://github.com/brettwooldridge/HikariCP/blob/dev/src/main/java/com/zaxxer/hikari/HikariConfig.java#L56

mosabua commented:
Will you be able to rebase this PR and address comments @arturmkr ?

arturmkr commented:
@mosabua 
I have merged the latest main into this branch to resolve conflicts and make sure it works with the current codebase.

Since this PR has been open for about 4 months, I would really appreciate it if someone could take a look at it.

Thanks!

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/persistence/TestJdbcConnectionManagerPool.java:
```diff
@@ -0,0 +1,175 @@
+final class TestJdbcConnectionManagerPool
+{
+    @Test
+    void blocksWhenExceedingMaxPoolSize()
```
https://trino.io/docs/current/develop/tests.html#conventions-and-recommendations 
* Test method must start with test, for example testExplain()

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/ha/persistence/JdbcConnectionManager.java:
```diff
@@ -114,4 +129,44 @@ private void startCleanUps()
+                return existing;
+            }
+
+            HikariConfig cfg = new HikariConfig();
```
Could you avoid abbreviations? https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#avoid-abbreviations

arturmkr replied:
Fixed

mosabua reviewed (approved):
Looks good to me. Any concerns with merging @ebyhr and @trinodb/maintainers-gateway ?

mosabua commented:
@arturmkr can you address the feedback from @ebyhr so we can maybe get this into the 19 release

Chaho12 commented:
@arturmkr Let us know when you are ready :)

arturmkr committed:
Add support for connection pool with HikariCP

arturmkr commented:
Hi @Chaho12 @ebyhr 

Sorry for the long delay in following up on this PR and thank you for your patience and for the detailed review feedback.

I have now updated the branch and addressed all the actionable review comments. While revisiting the implementation, I also realized that the original change only applied connection pooling to routing-group database connections, while the gateway metadata, query history, and scheduled query-history cleanup were still using non-pooled connections.

The connection pool is now used consistently for all gateway datastore access. I’ve also added coverage for pool reuse, pool limits, invalid configuration, shutdown, and the gateway metadata/query-history paths.

I merged the latest main, squashed the changes into a single commit, and ran the complete gateway-ha test suite: 376 tests passed with no failures or errors.

Could you please take another look when you have a chance? 

Thanks again!

oh0873 commented:
Hi @arturmkr, would you be open to also making `keepaliveTime` and `maxLifetime` configurable?

arturmkr commented:
> Hi @arturmkr, would you be open to also making `keepaliveTime` and `maxLifetime` configurable?

Done! Both `keepaliveTime` and `maxLifetime` are now configurable as optional data store settings. HikariCP defaults are preserved when they are not specified. I also added configuration tests and updated the documentation.

arturmkr commented:
@mosabua , can we proceed with merging the PR? are there any blockers?
